### PR TITLE
Feature complex schema 2

### PR DESCRIPTION
**Pull request by nbgl**
This is a pull request with the same changes as #54 but a cleaner commit history.

### CODE REVIEW

**Comment by nbgl** [comment]
Thanks for noticing the missing unit tests for unicode reader. I'll fix that.

I share your concerns about the defaults handling, but I can't think of a better way to do it. You mentioned using JsonSchema to automatically fill in the default values, but it doesn't seem to be a very elegant solution. The JsonSchema people themselves seem to be against it (see http://python-jsonschema.readthedocs.io/en/latest/faq/#why-doesn-t-my-schema-s-default-property-set-the-default-on-my-instance). Let's have a chat about potential ways of improving this.

File: clkhash/cli.py [hunk 1]
```diff
@@ -271,7 +272,23 @@ def generate(size, output, schema):
...
+
+
+@cli.command('generate-default-schema',
+             short_help='get the default schema used in generated random PII')
+@click.argument('output', type=click.Path(writable=True,
+                                          readable=False,
+                                          resolve_path=True))
+def generate_default_schema(output):
+    """Get default schema for fake PII"""
+    original_path = os.path.join(os.path.dirname(__file__),
```
Comment (hardbyte) [inline clkhash/cli.py]: This will need to use `pkgutil.get_data`

Comment (nbgl) [reply clkhash/cli.py]: Thanks! This is fixed now.

File: clkhash/schema.py [hunk 2]
```diff
@@ -1,36 +1,295 @@
...
+
+    try:
+        schema_bytes = pkgutil.get_data('clkhash',
+            'master-schemas/{}'.format(file_name))
+    except IOError as e:  # In Python 3 we can be more specific with
+                          # FileNotFoundError, but that doesn't exist in
+                          # Python 2.
+        msg = ('The master schema could not be found. The schema cannot be '
+               'validated. Please file a bug report.')
+        raise_from(MasterSchemaError(msg), e)
```
Comment (hardbyte) [inline clkhash/schema.py]: I've not come across `raise_from` before - very cool!

File: clkhash/schema.py [hunk 3]
```diff
@@ -1,36 +1,295 @@
...
+    
+    if schema_bytes is None:
+        msg = ('The master schema could not be loaded. The schema cannot be '
+               'validated. Please file a bug report.')
+        raise MasterSchemaError(msg)
+
+    return schema_bytes
+
+
+class SchemaError(Exception):
```
Comment (hardbyte) [inline clkhash/schema.py]: Move these exceptions to the top of the file (or to an exceptions module)

File: clkhash/schema.py [hunk 4]
```diff
@@ -1,36 +1,295 @@
...
+    except jsonschema.exceptions.ValidationError as e:
+        raise_from(SchemaError('The schema is not valid.'), e)
+    except jsonschema.exceptions.SchemaError as e:
+        msg = ('The master schema is not valid. The schema cannot be '
+               'validated. Please file a bug report.')
+        raise_from(MasterSchemaError(msg), e)
 
-    return schema
 
+def schema_from_json_file(schema_file):
```
Comment (hardbyte) [inline clkhash/schema.py]: Seems a bit inconsistent that this function isn't a classmethod but `Schema.from_json_dict` is, and that this doesn't allow the caller to not validate.

File: clkhash/backports.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import sys
+
+from mypy_extensions import Arg, DefaultNamedArg
+
+
+try:
+    int_from_bytes = int.from_bytes
+except AttributeError:
+    import codecs
+    def __int_from_bytes(bytes, byteorder, signed=False):
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/backports.py]: I had the same concern, but I used the name `bytes` for consistency with `int.from_bytes` in Python 3. See https://docs.python.org/3/library/stdtypes.html#int.from_bytes.

File: clkhash/schema.py (hunk 2 above)
Comment (nbgl) [reply clkhash/schema.py]: It’s a `future` backport of the `raise foo from bar` statement in Python 3. Lets you change the type of the exception—or add extra context—without losing `bar`’s traceback.

File: clkhash/backports.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        # type: (Sequence[int], str, bool) -> int
+        """ Emulate Python 3's int.from_bytes.
+
+            Kudos: https://stackoverflow.com/a/30403242 (with
+            modifications)
+
+            :param bytes_: The bytes to turn into an `int`.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/backports.py]: Thank you, fixed!

File: clkhash/cli.py [hunk 7]
```diff
@@ -271,7 +272,23 @@ def generate(size, output, schema):
...
-    randomnames.save_csv(pii_data.names, pii_data.schema, output)
+    randomnames.save_csv(
+        pii_data.names,
+        [f.identifier for f in pii_data.SCHEMA.fields],
+        output)
+
+
+@cli.command('generate-default-schema',
+             short_help='get the default schema used in generated random PII')
+@click.argument('output', type=click.Path(writable=True,
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/cli.py]: I need the path so I can use the operating system to perform the file copy, instead of reading and writing things manually. If you think it’d be helpful, I can subclass `click.Path` to implement this functionality.

File: clkhash/bloomfilter.py [hunk 8]
```diff
@@ -95,30 +102,41 @@ def double_hash_encode_ngrams_non_singular(ngrams,          # type: Iterable[str
...
+            bf[gi] = True
     return bf
 
 
 def blake_encode_ngrams(ngrams,          # type: Iterable[str]
-                       key,              # type: bytes
-                       k,                # type: int
-                       l                 # type: int
-                       ):
+                        keys,            # type: Sequence[bytes]
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/bloomfilter.py]: I wanted to have a common type signature for all the hashing functions, so I could access them in a uniform way. (More specifically, it made sense to have a function that takes a hash function name from the schema and returns a hashing function with a particular signature.)

File: clkhash/bloomfilter.py [hunk 9]
```diff
@@ -255,55 +291,57 @@ def fold_xor(bloomfilter,  # type: bitarray
...
-                        keys,         # type: Tuple[Tuple[bytes, ...]]
-                        xor_folds=0,  # type: int
-                        l=1024,       # type: int
-                        k=30,         # type: int
-                        ngram_encoding=NgramEncodings.DOUBLE_HASH
+def crypto_bloom_filter(record,          # type: Sequence[Text]
+                        tokenizers,      # type: List[Callable[[Text], Iterable[Text]]]
+                        field_hashing,   # type: List[field_formats.FieldHashingProperties]
+                        keys,            # type: Sequence[Sequence[bytes]]
+                        hash_properties  # type: clkhash.schema.GlobalHashingProperties
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/bloomfilter.py]: Yes, you’re right. It made sense at first to have `schema` as part of the `clkhash` namespace because ‘schema’ is also a common noun. But it got messy really quickly.

File: clkhash/field_formats.py [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+""" Classes that specify the requirements for each column in a dataset.
+    They take care of validation, and produce the settings required to
+    perform the hashing.
+"""
+
+from __future__ import unicode_literals
+
+import abc
+from datetime import datetime
+import sre_constants
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Sure. Thanks for fixing that!

File: clkhash/field_formats.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+    pass
+
+
+class InvalidSchemaError(ValueError):
+    """ The schema is not valid.
+
+        This exception is raised if, for example, a regular expression
+        included in the schema is not syntactically correct.
+    """
+    # TODO: Consider getting rid of this and just using ValueError
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Sure thing!

File: clkhash/field_formats.py [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+    def from_json_dict(cls, json_dict):
+        # type: (Dict[str, Any]) -> FieldHashingProperties
+        """ Make a HashingProperties object from a dictionary.
+
+            The dictionary must have have an 'ngram' key. It may have
+            'positional' and 'weight' keys; if these are missing, then
+            they are filled with the default values. The encoding is
+            always set to the default value.
+
+            :param hash_properties: The dictionary to use.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: It’s useful to know that the dictionary is restricted to JSON data types.

File: clkhash/schema.py [hunk 13]
```diff
@@ -1,36 +1,295 @@
...
-        ]
+import pkgutil
+from typing import Any, Dict, Hashable, List, Optional, Text, TextIO, Tuple
+
+from future.utils import raise_from
+import jsonschema
+
+from clkhash import field_formats
+
+# These are relative to this file. Using tuples to represent hierarches
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Indeed, thank you for removing it.

File: clkhash/schema.py [hunk 14]
```diff
@@ -1,36 +1,295 @@
...
+
+    @classmethod
+    def from_json_dict(cls, properties_dict):
+        # type: (Dict[str, Any]) -> GlobalHashingProperties
+        """ Make a GlobalHashingProperties object from a dictionary.
+
+            The dictionary must have a `'type'` key and a `'config'`
+            key. The `'config'` key must map to a dictionary
+            containinng a `'kdf'` key, which itself maps to a
+            dictioanry. That dictionary must have `'type'`, `'hash'`,
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Oops.

File: clkhash/schema.py [hunk 15]
```diff
@@ -1,36 +1,295 @@
...
+            `'keySize'`, `'salt'`, and `'type'` keys.
+
+            :param properties_dict: The dictionary to use.
+            :return: The resulting GlobalHashingProperties object.
+        """
+        result = cls()
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: The `__init__` method is there for convenience. In this particular case, I thought assigning directly to the attributes made for cleaner code.

File: clkhash/schema.py (hunk 3 above)
Comment (nbgl) [reply clkhash/schema.py]: I’ve moved them to the top of the file for now. Do you think it’d be worth it to have a separate exceptions module?

File: clkhash/schema.py (hunk 4 above)
Comment (nbgl) [reply clkhash/schema.py]: I agree, this should be a classmethod.

File: clkhash/schema.py [hunk 16]
```diff
@@ -1,36 +1,295 @@
...
+    __slots__ = ('k', 'l', 'xor_folds', 'hash_type',
+                 'hash_prevent_singularity', 'kdf_type', 'kdf_hash',
+                 'kdf_info', 'kdf_salt', 'kdf_key_size')
+
+    def __init__(self, **kwargs):
+        # type: (...) -> None
+        """ Make a GlobalHashingProperties object from keyword
+            arguments.
+
+            :param k: Value of `self.k`.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Sure, I’ll get rid of these.

File: clkhash/bloomfilter.py [hunk 17]
```diff
@@ -95,30 +102,41 @@ def double_hash_encode_ngrams_non_singular(ngrams,          # type: Iterable[str
...
     bf.setall(False)
     for m in ngrams:
-        sha1hm = int(hmac.new(key_sha1, m.encode(), sha1).hexdigest(), 16) % l
-        md5hm = int(hmac.new(key_md5, m.encode(), md5).hexdigest(), 16) % l
+        m_bytes = m.encode(encoding=encoding)
+
+        sha1hm_bytes = hmac.new(key_sha1, m_bytes, sha1).digest()
+        md5hm_bytes = hmac.new(key_md5, m_bytes, md5).digest()
+
+        sha1hm = int_from_bytes(sha1hm_bytes, 'big') % l
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/bloomfilter.py]: Okay, I’ll avoid this in the future.

File: clkhash/field_formats.py [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+        """ Make a FieldHashingProperties object from keyword
+            arguments.
+
+            :param encoding: (optional) Value of `self.encoding`.
+            :param ngram: (optional) Value of `self.ngram`.
+            :param positional: (optional) Value of `self.positional`.
+            :param weight: (optional) Value of `self.weight`.
+        """
+        if 'encoding' in kwargs:
+            self.encoding = kwargs['encoding']
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: The `__init__` method as it stands is mostly for convenience, as a prettier alternative to assigning to attributes. I agree that there should be a better way to make a schema in the API. However, we are overdue for an API review (#26), so it may be better to wait until then and consider it in the broader context of all of Clkhash.

File: clkhash/field_formats.py [hunk 19]
```diff
@@ -0,0 +1,544 @@
...
+        if 'ngram' in kwargs:
+            self.ngram = kwargs['ngram']
+        if 'positional' in kwargs:
+            self.positional = kwargs['positional']
+        if 'weight' in kwargs:
+            self.weight = kwargs['weight']
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Same as above. Let’s consider this along with #26.

File: clkhash/field_formats.py [hunk 20]
```diff
@@ -0,0 +1,544 @@
...
+        :ivar weight: Controls the weight of the field in the Bloom
+            filter.
+    """
+    __slots__ = ('encoding', 'ngram', 'positional', 'weight')
+
+    DEFAULT_ENCODING = 'utf-8'
+    DEFAULT_POSITIONAL = False
+    DEFAULT_WEIGHT = 1
+
+    def __init__(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I wanted to be able to pass the unused variables straight to the superclass initialiser. Maybe this is not the best way of doing this. This is something we can talk about.

File: clkhash/schema.py [hunk 21]
```diff
@@ -1,36 +1,295 @@
...
+            of `HKDFconfig` for details.
+        :param kdf_key_size: The size of the derived keys in bytes.
+
+    """
+    __slots__ = ('k', 'l', 'xor_folds', 'hash_type',
+                 'hash_prevent_singularity', 'kdf_type', 'kdf_hash',
+                 'kdf_info', 'kdf_salt', 'kdf_key_size')
+
+    def __init__(self, **kwargs):
+        # type: (...) -> None
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: How do you use kwargs but also annotate types?

File: clkhash/schema.py [hunk 22]
```diff
@@ -1,36 +1,295 @@
...
+        if 'hash_type' in kwargs:
+            self.hash_type = kwargs['hash_type']
+        if 'hash_prevent_singularity' in kwargs:
+            self.hash_prevent_singularity = kwargs['hash_prevent_singularity']
+
+        if 'kdf_type' in kwargs:
+            self.kdf_type = kwargs['kdf_type']
+        if 'kdf_hash' in kwargs:
+            self.kdf_hash = kwargs['kdf_hash']
+        if 'kdf_info' in kwargs:
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: `kdf_info` is used in key derivation (I don’t really know what its purpose there is).

File: clkhash/bloomfilter.py [hunk 23]
```diff
@@ -3,31 +3,37 @@
...
+    # Ignore because otherwise Mypy raises errors, thinking that
+    # blake2b is already defined.
 
 def double_hash_encode_ngrams(ngrams,          # type: Iterable[str]
-                              keys,            # type: Tuple[bytes, ...]
+                              keys,            # type: Sequence[bytes]
                               k,               # type: int
-                              l                # type: int
+                              l,               # type: int
+                              encoding         # type: str
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/bloomfilter.py]: Good catch!

File: clkhash/schema.py [hunk 24]
```diff
@@ -1,36 +1,295 @@
...
+            'doubleHash' and 'blakeHash'.
+        :param hash_prevent_singularity: Ignored unless hash_type is
+            'doubleHash'. Prevents bloom filter collisions in certain
+            cases when True.
+        :param kdf_type: The key derivation function to use. Currently,
+            the only permitted value is 'HKDF'.
+        :param kdf_hash: The hash function to use in key derivation. The
+            options are 'SHA256' and 'SHA512'.
+        :param kdf_info: The info for key derivation. See documentation
+            of `HKDFconfig` for details.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: I didn’t realise! I’ll keep that in mind for the future.

File: clkhash/schema.py [hunk 25]
```diff
@@ -1,36 +1,295 @@
...
+        # type: (Dict[str, Any], bool) -> Schema
+        """ Make a Schema object from a dictionary.
+
+            The dictionary must have a `'features'` key specifying the
+            columns of the dataset. It must have a `'version'` key
+            containing the master schema version that this schema
+            conforms to. It must have a `'hash'` key with all the
+            globals.
+
+            :param schema_dict: The dictionary to use.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Sure, I’ve fixed that.

File: clkhash/schema.py [hunk 26]
```diff
@@ -1,36 +1,295 @@
...
+        # type: (...) -> None
+        """ Make a Schema object from keyword arguments.
+
+            :param version: (optional) Value of `self.version`.
+            :param global_properties: (optional) Value of
+                `self.global_properties`.
+            :param fields: (optional) Value of `self.fields`.
+        """
+        if 'version' in kwargs:
+             self.version = kwargs['version']
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: As above, I think this is something we should tackle with #26.

File: clkhash/schema.py [hunk 27]
```diff
@@ -1,36 +1,295 @@
...
+            :param schema_dict: The dictionary to use.
+            :param validate: (default True) Should we throw if the
+                schema does not conform to the master schema?
+            :return: The resulting Schema object.
+        """
+        if validate:
+            # This raises iff the schema is invalid.
+            validate_schema_dict(schema_dict)
+
+        result = cls()
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: As before, I think this is something we should tackle with #26.

File: clkhash/schema.py [hunk 28]
```diff
@@ -1,36 +1,295 @@
...
+
+class Schema(object):
+    """ Overall schema for hashing.
+
+        A single object contains all the settings you will need, want,
+        or desire.
+
+        :ivar version: Version for the schema. Needed to keep behaviour
+            consistent between Clkhash versions for the same schema.
+        :ivar hashing_globals: Hashing globals. Your salts and lengths
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Thanks for fixing that docstring.

File: clkhash/schema.py [hunk 29]
```diff
@@ -1,36 +1,295 @@
...
+        result.kdf_info = base64.b64decode(
+            properties_dict['kdf']['info'])
+        result.kdf_salt = base64.b64decode(
+            properties_dict['kdf']['salt'])
+        result.kdf_key_size = properties_dict['kdf']['keySize']
+
+        return result
+
+
+class Schema(object):
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/schema.py]: Thanks for that. I’ve exposed `clkhash.schema` and `clkhash.field_formats`.

File: docs/index.rst [hunk 30]
```diff
@@ -12,6 +12,7 @@ Welcome to CLK hash's documentation!
 
    clkhash
    schema
+   field_formats
```
Comment (hardbyte) [inline docs/index.rst]: I'm surprised that `schema.rst` didn't need updating. However given the large doc changes in `develop` recently it would be best to deal with after merging.

**Comment by hardbyte** [comment]
My changes are [here](https://github.com/n1analytics/clkhash/compare/feature-complex-schema-rebased...feature-complex-schema-review-2?expand=1). Please ask for another review after bringing up to date with `develop`.

File: tests/test_e2e_hashing.py [hunk 31]
```diff
@@ -52,13 +52,13 @@ def test_different_weights(self):
                 StringSpec(
                     identifier='some info',
                     hashing_properties=FieldHashingProperties(
-                        encoding=FieldHashingProperties.DEFAULT_ENCODING,
+                        encoding=FieldHashingProperties._DEFAULT_ENCODING,
```
[the remark this answers is not in the document]
Comment (nbgl) [reply tests/test_e2e_hashing.py]: These tests already exist in test_field_formats.py.

File: clkhash/field_formats.py [hunk 32]
```diff
@@ -0,0 +1,515 @@
...
+        included in the schema is not syntactically correct.
+    """
+
+
+class FieldHashingProperties(object):
+    """ Stores the information needed to to create a hash from each
+        entry of a type.
+
+        :ivar encoding: The encoding to use when converting the string
+            to bytes. Refer to Python's documentation for possible
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I’ve included the link.

File: clkhash/field_formats.py [hunk 33]
```diff
@@ -0,0 +1,515 @@
...
+class InvalidSchemaError(ValueError):
+    """ The schema is not valid.
+
+        This exception is raised if, for example, a regular expression
+        included in the schema is not syntactically correct.
+    """
+
+
+class FieldHashingProperties(object):
+    """ Stores the information needed to to create a hash from each
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I’ve updated it to be more clear.

File: clkhash/field_formats.py [hunk 34]
```diff
@@ -0,0 +1,515 @@
...
+        entry of a type.
+
+        :ivar encoding: The encoding to use when converting the string
+            to bytes. Refer to Python's documentation for possible
+            values.
+        :ivar ngram: The n in n-gram. Possible values are 0, 1, and 2.
+        :ivar positional: Controls whether the n-grams are positional.
+        :ivar weight: Controls the weight of the field in the Bloom
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: As we’ve discussed earlier, I’ve made the assumption that users will be using `from_json_dict` to make these classes. I am happy to talk about making it easier to do it through `__init__`, but let’s worry about that in #26.

File: clkhash/field_formats.py [hunk 35]
```diff
@@ -0,0 +1,515 @@
...
+            self.hashing_properties = kwargs['hashing_properties']
+
+    @classmethod
+    def from_json_dict(cls, field_dict):
+        # type: (Dict[str, Any]) -> FieldSpec
+        """ Initialise a FieldSpec object from a dictionary of
+            properties.
+
+            This dictionary must contain a `'hashing'` key that meets
+            the requirements of :meth:`HashingProperties.__init__()`.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Done!

File: clkhash/field_formats.py [hunk 36]
```diff
@@ -0,0 +1,515 @@
...
+            contents are passed to HashingProperties.__init__.
+
+            :param properties: The properties dictionary.
+        """
+        result = cast(IntegerSpec,  # For Mypy.
+                      super().from_json_dict(json_dict))
+
+        format_ = json_dict['format']
+
+        result.minimum = format_.get('minimum', 0)  # No integers < 0.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: As we’ve discussed earlier, I’ve made the assumption that users will be using `from_json_dict` to make these classes (let’s worry about making it easier to do it through `__init__` in #26). Negative integers are not allowed by according to the master schema.

File: clkhash/field_formats.py [hunk 37]
```diff
@@ -0,0 +1,515 @@
...
+    def from_json_dict(cls, json_dict):
+        # type: (Dict[str, Any]) -> IntegerSpec
+        """ Make a IntegerSpec object from a dictionary containing its
+            properties.
+
+            The dictionary may contain `'minimum'` and `'maximum'`
+            keys. In addition, it must contain a `'hashing'` key, whose
+            contents are passed to HashingProperties.__init__.
+
+            :param properties: The properties dictionary.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Thank you, fixed.

File: clkhash/field_formats.py [hunk 38]
```diff
@@ -0,0 +1,515 @@
...
+            result.min_length = format_.get('minLength')
+            result.max_length = format_.get('maxLength')
+
+        return result
+
+    def validate(self, str_in):
+        # type: (Text) -> None
+        """ Validates an entry in the field.
+
+            Raises `InvalidEntryError` iff the entry is invalid.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I think this is very important to make clear. The function doesn’t return anything; rather, it’s entire behaviour is to raise an exception if things are invalid. I’d rather not relegate this information down the bottom, where it is likely to be missed.

File: clkhash/field_formats.py [hunk 39]
```diff
@@ -0,0 +1,515 @@
...
+        super().validate(str_in)  # Validate encoding.
+
+        if hasattr(self, 'regex'):
+            match = self.regex.match(str_in)
+            if match is None:
+                raise InvalidEntryError(
+                    'Expected entry that conforms to regular expression '
+                    "'{}'. Read '{}'.".format(self.regex.pattern, str_in))
+
+        else:
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: That’s true. I didn’t have strong feelings about this, so I made the same decision as the master schema does and made them the same type. This can be changed if you’d like.

File: clkhash/field_formats.py [hunk 40]
```diff
@@ -0,0 +1,515 @@
...
+            encoding cannot represent the string.
+
+            :param str_in: String to validate.
+            :raises InvalidEntryError: When entry is invalid.
+            :raises ValueError: When self.case is not one of the
+                permitted values (`'lower'`, `'upper'`, or `'mixed'`).
+        """
+        super().validate(str_in)  # Validate encoding.
+
+        if hasattr(self, 'regex'):
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I can’t think of a much better way of doing this. I can set `regex` to `None` if it is undefined. Splitting `StringSpec` into multiple classes will be inconsistent with the master schema, although it would certainly solve this problem. Alternatively, we can modify the master schema.

File: clkhash/field_formats.py [hunk 41]
```diff
@@ -0,0 +1,515 @@
...
+            specification is not regex-based.
+        :ivar max_length: The maximum length of the string. `None` if
+            there is no maximum length. Present only if the
+            specification is not regex-based.
+    """
+    _DEFAULT_CASE = 'mixed'
+
+    def __init__(self, **kwargs):
+        # type(...) -> None
+        """ Make a HashingProperties object from keyword arguments.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I’ve made the assumption that these classes will be created using `from_json_dict`. Let’s relegate this to #26.

File: clkhash/field_formats.py [hunk 42]
```diff
@@ -0,0 +1,515 @@
...
+        in the dataset. Subclasses validate entries, and modify the
+        hashing_properties ivar to customise hashing procedures.
+
+        :ivar identifier: The name of the field.
+        :ivar description: Description of the field format.
+        :ivar hashing_properties: The properties for hashing.
+    """
+    def __init__(self, **kwargs):
+        # type: (...) -> None
+        """ Make a HashingProperties object from keyword arguments.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I’ve changed the docstring to make it clear that all it does it set the object’s attributes to values specified in the keyword arguments.

You raise a good point about including types in the docstrings. I’ll add those.

File: clkhash/field_formats.py [hunk 43]
```diff
@@ -0,0 +1,515 @@
...
+            self.regex = kwargs['regex']
+        if 'case' in kwargs:
+            self.case = kwargs['case']
+        if 'min_length' in kwargs:
+            self.min_length = kwargs['min_length']
+        if 'max_length' in kwargs:
+            self.max_length = kwargs['max_length']
+
+    @classmethod
+    def from_json_dict(self, json_dict):
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Indeed, I missed that. I’ve changed it to be `cls`.

File: clkhash/field_formats.py [hunk 44]
```diff
@@ -0,0 +1,515 @@
...
+class StringSpec(FieldSpec):
+    """ Represents a field that holds strings.
+
+        One way to specify the format of the entries is to provide a
+        regular expression that they must conform to. Another is to
+        provide zero or more of: minimum length, maximum length, casing
+        (lower, upper, mixed).
+
+        Each string field also specifies an encoding used when turning
+        characters into bytes.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: The encoding is stored in `StringSpec.hashing_properties` as it is required for hashing.

File: clkhash/field_formats.py [hunk 45]
```diff
@@ -0,0 +1,515 @@
...
+            An entry is invalid iff (1) a pattern is part of the
+            specification of the field and the string does not match
+            it; (2) the string does not match the provided casing,
+            minimum length, or maximum length; or (3) the specified
+            encoding cannot represent the string.
+
+            :param str_in: String to validate.
+            :raises InvalidEntryError: When entry is invalid.
+            :raises ValueError: When self.case is not one of the
+                permitted values (`'lower'`, `'upper'`, or `'mixed'`).
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I’ve made the assumption that (for now) these classes would be created with `from_json_dict` instead of `__init__`. The validation ensures that the parameters are correct, ValueError should really not be raised. It may make more sense to remove it from the docstring.

File: clkhash/field_formats.py [hunk 46]
```diff
@@ -0,0 +1,515 @@
...
+            with a Python-conformant encoding. It must also contain a
+            `'hashing'` key, whose contents are passed to
+            HashingProperties.__init__.
+
+            Possible keys also include `'pattern'`, `'case'`,
+            `'minLength'`, and `'maxLength'`.
+
+            :param properties: The properties dictionary.
+            :raises InvalidSchemaError: When a regular expression is
+                provided but is not a valid pattern.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: I think it makes sense for this to be an `InvalidSchemaError`. This is the error we throw when the schema is invalid in other ways, such as if it doesn’t pass JsonSchema validation.

File: clkhash/field_formats.py [hunk 47]
```diff
@@ -0,0 +1,515 @@
...
+    """ Represents a field that holds integers.
+
+        Minimum and maximum values may be specified.
+
+        :ivar minimum: The minimum permitted value.
+        :ivar maximum: The maximum permitted value or None.
+    """
+    def __init__(self, **kwargs):
+        # type: (...) -> None
+        """ Make a HashingProperties object from keyword arguments.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Thanks, fixed.

File: clkhash/field_formats.py [hunk 48]
```diff
@@ -0,0 +1,515 @@
...
+        """
+        super().validate(str_in)
+
+        try:
+            value = int(str_in, base=10)
+        except ValueError as e:
+            msg = 'Invalid integer. Read {}.'.format(str_in)
+            raise_from(InvalidEntryError(msg), e)
+
+        if value < self.minimum:
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: The minimum is always defined, according to the master schema. (It has a default of 0.)

File: clkhash/field_formats.py [hunk 49]
```diff
@@ -0,0 +1,515 @@
...
+        supported format is RFC3339.
+
+        :ivar format: The format of the date.
+    """
+    RFC3339_REGEX = re_compile_full(r'\d\d\d\d-\d\d-\d\d')
+    RFC3339_FORMAT = '%Y-%m-%d'
+
+    def __init__(self, **kwargs):
+        # type: (...) -> None
+        """ Make a HashingProperties object from keyword arguments.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Thanks

File: clkhash/field_formats.py [hunk 50]
```diff
@@ -0,0 +1,515 @@
...
+        """
+        super().__init__(**kwargs)
+
+        if 'format' in kwargs:
+            self.format = kwargs['format']
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: The master schema doesn’t specify a default value for this. Rather, it specifies a list of permitted values in a mandatory field—this list has length 1.

File: clkhash/field_formats.py [hunk 51]
```diff
@@ -0,0 +1,515 @@
...
+
+            Raises `InvalidEntryError` iff the entry is invalid.
+
+            An entry is invalid iff (1) the string does not represent a
+            date in the correct format; or (2) the date it represents
+            is invalid (such as 30 February).
+
+            :param str_in: String to validate.
+            :raises InvalidEntryError: When entry is invalid.
+            :raises ValueError: When self.format is unrecognised.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: This condition is already validated when the object is being created using `Schema.from_json_dict`. So this is more of a runtime sanity check and the exception should never be raised. Would it be better to remove it from the docstring?

File: clkhash/field_formats.py [hunk 52]
```diff
@@ -0,0 +1,515 @@
...
+            msg = ('Expected integer value of at most {}. Read {}.'
+                   .format(self.maximum, value))
+            raise InvalidEntryError(msg)
+
+
+class DateSpec(FieldSpec):
+    """ Represents a field that holds dates.
+
+        A format for the date may be specified. Currently, the only
+        supported format is RFC3339.
```
[the remark this answers is not in the document]
Comment (nbgl) [reply clkhash/field_formats.py]: Done!

**Comment by nbgl** [comment]
Thanks for that review, Brian. I’ve worked on some of your concerns regarding initialisation of `FieldSpec` subclasses with `__init__`. However, I’d like to relegate most of this work to #26 since the entire Clkhash API is overdue for a review and I’d like to consider all these questions in the broader context.
